tecton_client/_internal/tecton_client.py

```diff
@@ -0,0 +1,102 @@
+import json
+from typing import Any, Dict, Optional, Union
+from urllib.parse import urljoin
+
+import httpx
+from httpx import HTTPStatusError
+
+from tecton_client._internal.data_types import GetFeatureServiceMetadataResponse, GetFeaturesResponse
+from tecton_client._internal.utils import (
+    build_get_feature_service_metadata_request,
+    build_get_features_request,
+    get_default_headers,
+    validate_request_args,
+)
+from tecton_client.exceptions import convert_exception
+
+
+class TectonClient:
+    """A lightweight http client for interacting with features in Tecton. For the full sdk, use tecton-sdk"""
+
+    def __init__(
+        self, url: str, api_key: str, default_workspace_name: Optional[str] = None, client: httpx.Client = None
```

Comment:
Can we make this a dataclass:

We can use `__post_init__`: https://docs.python.org/3.11/library/dataclasses.html#init-only-variables

Reply:
The Client? The Client isn't a dataclass though. Dataclasses are things that are primarily meant as bags-of-data. You can add extra methods/properties to them, but they are still meant as bags of data. The client _isn't_ just a bag of data. The 'str' function on a Client probably shouldn't print out it's properties, and two clients with the same init values (url, api_key, workspace_name) shouldn't be equal to each other in the way a dataclass would be.

Reply:
I'll leave it up to your judgement. We follow this pattern elsewhere in the main Tecton codebase.

Reasoning: I like it since it reduces boilerplate / clearer class definition (it is readily apparent the attributes + types of the object). You can set `eq=False` and `repr=False` to avoid the creation of the `__repr__` and `__eq__` funcs and avoid these properties you mention.

Reply:
Yeah I think I won't in this case. I think we'd end up disabling more of the dataclass functionality than we'd get, and the way that the api_key would need to be a private variable even though it's passed in as an argument, etc, just doesn't seem like it fits with dataclass,